rasa/core/policies/policy.py

```diff
@@ -215,6 +215,31 @@ def predict_action_probabilities(
 
         raise NotImplementedError("Policy must have the capacity to predict.")
 
+    def predict_action_probabilities_with_diagnostics(
+        self,
+        tracker: DialogueStateTracker,
+        domain: Domain,
+        interpreter: NaturalLanguageInterpreter,
+        **kwargs: Any,
+    ) -> Tuple[List[float], Optional[Dict[Text, Any]]]:
+        """Predicts the next action and diagnostic data for debugging.
+
+        Args:
+            tracker: the :class:`rasa.core.trackers.DialogueStateTracker`
+            domain: the :class:`rasa.shared.core.domain.Domain`
+            interpreter: Interpreter which may be used by the policies to create
+                additional features.
+
+        Returns:
+             The list of probabilities for the next actions and a dictionary of
+             diagnostic data (or None).
+        """
+
+        return (
+            self.predict_action_probabilities(tracker, domain, interpreter, **kwargs),
+            None,
+        )
```

Comment:
@wochinge @tmbo what do you think about adding this method? Or should we rather update `self.predict_action_probabilities` to always return some diagnostics, and then ignore it in default processor?

Reply:
One thing to point out; it might be hard to predict right now what sort of things we might want to add to the diagnosis in the future. I could imagine for DIET that I'd also like to retreive the __CLS__ embeddings. With that in mind there's something to be said to keep the two methods seperate. That way we might seperate concerns and never contaminate the method that is "mission-critical".

Reply:
@JEM-Mosig @koaning Why do we need to do a prediction in order get the model weights? Couldn't we also inspect the checkpoints files (e.g. using [this](https://www.tensorflow.org/guide/checkpoint#manually_inspecting_checkpoints))?

Reply:
@wochinge We are not interested in the attention weights during training, but only during inference. Attention weights are not model weights, but tensors that depend on the input (kind of an intermediate result of the prediction process). So to compute them, we have to run much of the usual prediction code anyway and we will also always need the end result of the prediction to interpret the attention weights.

Reply:
@wochinge we hope to be able to answer questions like "How will DIET respond to examples that it's never seen before?" and "How will DIET respond when we give it something that's very much unlike anything that it has seen before?". To answer these questions nicely we need to have some level of interactivity after the model is trained.

Reply:
Thanks for the explanation @JEM-Mosig ! I was wondering about that but the TensorFlow documentation of the layers didn't provide me any quick help 🙈 

I am facing currently a similar problem regarding the end-to-end implementation. In this case I need the `Policy` prediction to return to additional values. My current idea would be to return an object instead of primitive values (`List[float]`). When doing that we could also make additional information like the attention weights an optional parameter of the prediction. This way we wouldn't need a second `predict` method and it would be more future proof. I'll have more information today / tomorrow when I start implementing (you'd need to do the same for the `Component` predictions then). What would you think of an approach like this?

Reply:
@wochinge This sounds way more solid! How about the DIET side of this?https://github.com/RasaHQ/rasa/blob/2501deacd8193e1830d5ffedf1523fc61a6d80b7/rasa/nlu/classifiers/diet_classifier.py#L891

Reply:
That's something which we (you) would need to do separately. I'm currently doing it for the policies and it's not too much work. Mostly adapting tests.

Reply:
you can have a glimpse here: https://github.com/RasaHQ/rasa/pull/7136/files

Reply:
@wochinge Since `Component`s just modify the given message and return `None`, I don't think introducing a new kind of object here makes sense. I think if we just change the signature of the `process` function to return the diagnostic data, then this data could be returned as a dict, as there is nothing else to return. I shouldn't have written the `process_with_diagnostics` function in the first place. 
Unless we want to change the return type of `batch_predict` from the dict to an object, but I don't think that's necessary, right?

Reply:
True, we can just add some more values to `Message` 👍 

> Unless we want to change the return type of batch_predict from the dict to an object, but I don't think that's necessary, right?

👍  It's probably too early to use an object there although the `Dict` is not ideal.

Reply:
Wait, I was thinking of _not_ adding diagnostic data to the messages, but instead returning it directly from the `process` function.

Reply:
@wochinge should I merge your PR branch into this one (for the new return objects)?

Reply:
what would be the advantage / pros/ cons of doing that? Returning it would require a change to the interface 😬

Reply:
@JEM-Mosig ?

Reply:
@wochinge Diagnostic data can involve large arrays which would clutter up the messages. Returning the diagnostic data directly is an interface change, but since nothing was returned before, this change is trivial (no other function expects any return value).